test_pull_requests

```diff
@@ -152,6 +185,115 @@ def submitted_tests_for_branch(branch)
   $submitted_tests[branch] ? $submitted_tests[branch] : $submitted_tests['*']
 end
 
+# Helper function to create appropriate regex for matching/extracting
+# :build_url field from bot comment
+def build_url_field_regex
+  if Properties &&
+    / \((#{Regexp::escape(Properties['jenkins_url'])}[^)]+)\)/
+  else
+    # Use this test to set the build url matching pattern
+    if ENV['BUILD_URL']
+      # Don't set build_url directly from the env var, since we want
+      # to make sure they match
+      / \((#{Regexp::escape(ENV['BUILD_URL'])})\)/
+    else
+      / \((https?:[^)]+)\)/
+    end
+  end
+end
+
+# Helper function to pull the test prefix from the settings object or
+# try to determine it from the provided bot_comment_body
+def extract_test_prefix(bot_comment_body, settings=nil)
+  if settings
+    settings['test_prefix']
+  else
+    # Do our best to find the prefix since we don't know it beforehand
+    if bot_comment_body =~ /^([^:(]+): /
+      $1.strip
+    elsif bot_comment_body =~ /^([^ (]+) /
+      $1.strip
+    end
+  end
+end
+
+# Pulls out as much metadata from the bot comment as it can, including
+# state (AKA content), Jenkins build URL, base repo commit ID, and the
+# extended tests metadata
+def extract_bot_comment_fields(bot_comment_body, settings=nil)
```

Comment:
> I meant more work in terms of the interpreter. 

I don't mind making robots do work :)

> But we're only passing a reference, we're not doing a deep copy

It isn't about the difficulty or complexity of the copy or pass, it's about the fact that the methods you wrote never need to see that data, never need to know it could possibly exist, never need to be understood in the context of it, etc. There's no good reason to proliferate a massive data structure like that.

> That shouldn't happen

Right, but the code as written makes no assertion about that but does make that assumption.

> Unpacking values that are tightly-coupled ... smells bad.

I disagree. Package them some other way if you want, if the current object isn't correctly reflecting the coupling in the data, but passing a massive structure holding information as diverse as team membership data and flake identification config into a function that formats a string or runs some regex is objectively wrong.